File: Python/63.Unique Paths II.py
#从左上角到右下角的不同路径数(只能向右向下),有'1'的位置是障碍
class Solution(object):
    def uniquePathsWithObstacles(self, obstacleGrid):
        m = len(obstacleGrid)
        if m==0:
            return 0
        n = len(obstacleGrid[0])
        g = [[0] * n for i in range(m)]
        for i in range(m):
            if obstacleGrid[i][0] == 1: break
            else: g[i][0] = 1
        for j in range(n):
            if obstacleGrid[0][j] == 1: break
            else: g[0][j] = 1
        for i in range(1, m):
            for j in range(1, n):
                if obstacleGrid[i][j]==0:
                    g[i][j] = g[i][j-1] + g[i-1][j]
                else:
                    g[i][j] = 0
        return g[m-1][n-1]

s = Solution()
print(s.uniquePathsWithObstacles([[0,0,0],[0,1,0],[0,0,0]]))
print(s.uniquePathsWithObstacles([[1,0]]))

File: C++/Array/63.Unique Paths II.cpp
/*
二维数组，中间有障碍1，空位置0，求从左上到右下有多少种路径（只能往下和往右移动）。
*/
/*
思路：左上开始，一个位置是1，则将它置为0，如果是0，则置为1。
后边的每个位置，如果是1（即障碍），则将该位置置为0，表示从前面到该位置的路径数是0
如果该位置是0，左加上。
 */
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
	int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
		if(obstacleGrid[0][0]==1)	return 0;
        for(int i=0; i < obstacleGrid.size(); ++i)
        {
        	for(int j=0; j < obstacleGrid[0].size(); ++j)
        	{
        		if(obstacleGrid[i][j]==0)
        		{
        			if(i==0 && j==0)	obstacleGrid[i][j] = 1;
        			if(j>0)	obstacleGrid[i][j] = obstacleGrid[i][j-1];
        			if(i>0)		obstacleGrid[i][j] += obstacleGrid[i-1][j];
        		}
        		else
        		{
        			obstacleGrid[i][j] = 0;
        		}
        	}
        }
        return obstacleGrid[obstacleGrid.size()-1][obstacleGrid[0].size()-1];
    }
};

int main()
{
	Solution s;
	vector<vector<int>> grid{{0,1,0},{0,1,0},{0,0,0}};
	cout<<s.uniquePathsWithObstacles(grid);
}

